### ISSUE TITLE
Resupport unfailable queries (v5.0.0-rc.4)

### ISSUE THREAD

**Issue by JordanMartinez** [slamdata/purescript-halogen#605]
In `v4.0.0`, the query API returned an `a`:
```purescript
query :: forall a. Query a -> HalogenM ... a Aff
```
In `v5.0.0-rc.4`, the query API now returns a `Maybe a`
```purescript
query :: forall a. Query a -> HalogenM ... (Maybe a) Aff
```
In v4, I could have my business logic (outside of Halogen component code) run a query that was guaranteed to never fail because it always returned an `a`. In v5, I can no longer do this. Rather, I have to handle a possible `Nothing` case, even though my code guarantees that that will never happen. I could use `unsafeJust`, but I'd rather not use partial functions if possible. 

The way I got around this was using `AVar`s in a "nested" sort of way. I don't think this is the cleanest and most understandable code that will be easy to maintain later, especially when using components with greater complexity. Still, my solution looked something like this:
```purescript
businessLogicCode :: Aff Value
businessLogicCode = do
  avar <- liftEffect AVar.empty
  -- a request is made on the root component
  void $ forkAff $ halogenIO.query $ H.request $ GetStuff prompt avar
  liftEffect $ AVar.take avar

rootComponentCode :: forall a. Query a -> Halogen ... (Maybe a) Aff
rootComponentCode = case _ of
  GetStuff prompt bizLogicAvar next -> do
    -- This comment represents code that does something with the `prompt` value...

    -- This block of code sets up the child component, which 
    -- actually calculates the result the business logic requires.
    -- Since it requires "cleaning up" the child component once that result is finished,
    -- it needs to create its own AVar for control flow. This is the "nested" idea I mentioned above.
    modify_ (\oldState -> updateState)
    childAvar <- liftEffect $ AVar.empty
    void $ forkAff $ H.query $ H.request $ CalculateValueUsingChild childAvar
    childResult <- liftEffect AVar.take childAvar
    modify_ (\oldState2 -> updateState2)

    -- Now that the child has been used and cleaned up,
    -- I'll pass the value back into the bizLogicAvar to continue executing
    -- my business logic.
    liftEffect $ AVar.put childResult bizLogicAvar
    pure $ Just next

childComponentCode :: forall a. Query a -> Halogen ... (Maybe a) Aff
childComponentCode = case _ of 
  CalculateValueUsingChild rootAvar next -> do
    -- does a lot of stuff in this component...
    -- but eventually the result gets calculated and passed back up
    -- into the root component.
    result <- computeResult
    void $ liftEffect $ AVar.put result rootAvar
    pure $ Just next
```

Is it possible to resupport queries that cannot fail in addition to those that can? Something like the following seems readable to me:
`H.failable $ H.request $ SomeFailableRequest`
`H.unfailable $ H.request $ SomeUnfailableRequest`

**Comment by garyb** [comment]
`H.query` always returned `Maybe`, because there's no ability to guarantee that querying a child component is going to work, since it might not exist.

The change in H5 was to support queries that can fail to `eval` as well, so you don't end up with a `Maybe (Maybe ...)` in situations where you want things to be failable (for example). This also makes proxy components more natural too, as they can directly pass a query to a child component and don't have to do things like raise a `Halt` if the child does not exist, etc. (in fact, `Halt` is gone entirely now)

**Comment by garyb** [comment]
Here's `query` in v4, as a reminder about its `Maybe`-ness: 

https://github.com/slamdata/purescript-halogen/blob/v4.0.0/src/Halogen/Query.purs#L94-L100

🙂

**Comment by garyb** [comment]
I think I see what's happened here, you're talking about the `query` that `DriverIO` gives you? This did indeed used to be `f ~> m`, but that's only possible when the top level component cannot be destroyed/disposed of.

**Comment by JordanMartinez** [comment]
Thanks for the quick reply. I didn't realize you'd have to deal with the
possibility of the child component not yet existing.

I won't be able to respond to this until much later today if not tomorrow.

**Comment by JordanMartinez** [comment]
> I think I see what's happened here, you're talking about the query that DriverIO gives you? This did indeed used to be f ~> m, but that's only possible when the top level component cannot be destroyed/disposed of.

Yeah, that is my focus. The idea is to implement the Onion Architecture. So, the business logic exists outside of the Halogen framework, so I can reuse it on different platforms (e.g. Node/C++ if I was using a terminal-based user interface, Electron/C++ if I choose to use a GUI on a desktop, Halogen/React if I was using a browser-based UI).

> This did indeed used to be f ~> m, but that's only possible when the top level component cannot be destroyed/disposed of.

Can it be returned back to `f ~> m`? Or does it need to be `f a -> m (maybe a)` in case a `HalogenIO.dispose` call is made before a `HalogenIO.query` call is made?

**Comment by garyb** [comment]
Yeah, that's exactly it - by having the ability to dispose of the top level component, it's the same situation as occurs for child components - if the `query` is made, and the component has been disposed, there's nothing available to evaluate the query.

**Comment by JordanMartinez** [comment]
I see. Thanks for clarifying that.

It seems like this reveals a flaw in my above workaround as well. If my code should ever dispose of the Halogen component before that query gets evaluated, then the `AVar.take` call would produce a deadlock because it will never return. Rather my program would just hang forever until I terminated it.

If I used `fromJust`, then a runtime error would be produced in that situation, which is a lot easier to debug than a deadlock. If such an error did occur, then I think the principle of "let it crash" is the only approach.
